returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,199 @@
+        if map_seq is None and map_seq_stream is None:
+            raise ValueError(f"{self}: need to either set map_seq or map_seq_stream")
+        if map_seq and map_seq_stream:
+            raise ValueError(f"{self}: cannot set both map_seq and map_seq_stream")
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq
+        self._map_seq_stream = map_seq_stream
+        self._map_outputs = map_outputs
+
+        self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self._estimated_num_seqs = self._dataset.estimated_num_seqs
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._dim_template_dict: Optional[TensorDict]
+
+        if self._map_outputs is not None:
+            self._dim_template_dict = TensorDict(self._map_outputs)
+        else:
+            # Some sub datasets need their seq order initialized for the shape/dtype accessor functions to work
+            self._dataset.init_seq_order(epoch=1)
+
+            def _make_tensor_template(name: str) -> Tensor:
+                feature_dims = [
+                    Dim(dimension=dim, name=f"{name}_dim{i + 1}")
+                    for i, dim in enumerate(self._dataset.get_data_shape(name))
+                ]
+                dims = [Dim(dimension=None, name=f"{name}_frame"), *feature_dims]
+                sparse_dim = None
+                if self._dataset.is_data_sparse(name):
+                    sparse_dim = Dim(dimension=self._dataset.get_data_dim(name), name=f"{name}_sparse")
+                    if name in self._dataset.labels:
+                        sparse_dim.vocab = Vocabulary.create_vocab_from_labels(self._dataset.labels[name])
+                dtype = self._dataset.get_data_dtype(name)
+                return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim)
+
+            tensor_dict = {
+                name: _make_tensor_template(name)
+                for name, num_output_def in self._dataset.num_outputs.items()
+                if isinstance(
+                    num_output_def, (tuple, list)
+                )  # Some datasets (e.g. OggZipDataset) have data that does not conform to the typings in num_outputs
+            }
+            self._dim_template_dict = TensorDict(tensor_dict)
+        self.num_outputs = {
+            k: (t.sparse_dim.size if t.sparse_dim else t.shape[-1] or 1, t.ndim)
+            for k, t in self._dim_template_dict.data.items()
+        }
+        self._default_input = "data" if "data" in self.num_outputs else next(iter(self.num_outputs.keys()))
+        self.num_inputs = self.num_outputs[self._default_input][0]
+
+        self.labels = {}
+        for k, t in self._dim_template_dict.data.items():
+            if t.vocab:
+                self.labels[k] = t.vocab.labels
+            elif t.sparse_dim:  # sparse_dim but not vocab
+                self.labels[k] = list(map(str, range(t.sparse_dim.dimension)))  # dummy labels
+
```

Comment:
You should also define most of the other accessor functions when possible (`get_data_shape`, `get_data_dtype`, etc).